### ISSUE TITLE
How to wait for loader to finish before dom object gets returned?

### ISSUE THREAD

**Issue by magician11** [jsdom/jsdom#1899]
### Basic info:

Node version is `7.9.0`
jsdom version is `11.0.0`

### Minimal reproduction case

```js
const { JSDOM } = require('jsdom');

JSDOM.fromURL('https://www.sfcinemacity.com/showtime/cinema/9936').then(dom => {
  const sfcinemaObject = dom.window.document.getElementsByClassName('.showtime-box');
  console.log(sfcinemaObject);
});
```

Results in `HTMLCollection {}`

Running `window.document.getElementsByClassName('showtime-box')` in the console in Chrome gives me
![screen shot 2017-06-27 at 3 51 26 pm](https://user-images.githubusercontent.com/3735849/27613498-a8978ba4-5b50-11e7-819c-c758b4e5db6a.png)

So in my code I get an empty set and in the Chrome browser I get an array of 4 elements.

I think this might be because in the browser there is a loader I see, and the code snippet is returning the dom object before that loading completes.

So is there a way to delay the result? Or something similar?

Thanks.

**Comment by domenic** [comment]
Try actually doing something with that HTMLCollection. It just logs differently in Node than in Chrome.

**Comment by Sebmaster** [comment]
Chances are you need the page to [execute JS](https://github.com/tmpvar/jsdom#executing-scripts) if it loads something dynamically.

**Comment by magician11** [comment]
@domenic HTMLCollection is an empty set. Looking for `li` elements by contrast in Node logs out

```
NodeList {
  '0': HTMLLIElement {},
  '1': HTMLLIElement {},
  '2': HTMLLIElement {},
  '3': HTMLLIElement {},
  '4': HTMLLIElement {},
  '5': HTMLLIElement {},
  '6': HTMLLIElement {},
  '7': HTMLLIElement {},
  '8': HTMLLIElement {},
  '9': HTMLLIElement {},
  '10': HTMLLIElement {},
  '11': HTMLLIElement {},
  '12': HTMLLIElement {},
  '13': HTMLLIElement {},
  '14': HTMLLIElement {},
  '15': HTMLLIElement {},
  '16': HTMLLIElement {},
  '17': HTMLLIElement {},
  '18': HTMLLIElement {},
  '19': HTMLLIElement {},
  '20': HTMLLIElement {},
  '21': HTMLLIElement {},
  '22': HTMLLIElement {},
  '23': HTMLLIElement {},
  '24': HTMLLIElement {},
  '25': HTMLLIElement {},
  '26': HTMLLIElement {},
  '27': HTMLLIElement {},
  '28': HTMLLIElement {},
  '29': HTMLLIElement {} }
```

@Sebmaster you mean like this:

```
const dom = new JSDOM(``, {
  url: 'https://www.sfcinemacity.com/showtime/cinema/9936',
  runScripts: 'dangerously'
});

const sfcinemaObject = dom.window.document.getElementsByClassName('.showtime-box');
console.log(sfcinemaObject);
```
?

This also returns an empty set.

**Comment by Sebmaster** [comment]
After you've enabled scripts, you'll also need to wait for the page to have fully loaded. Either by continually checking if the loader is still there, or you find some JS code on the page you can hook into.

**Comment by magician11** [comment]
@Sebmaster 

I tried simply adding a timeout before it referenced the dom

```
const dom = new JSDOM(``, {
  url: 'https://www.sfcinemacity.com/showtime/cinema/9936',
  runScripts: 'dangerously'
});

setTimeout(() => {
  const sfcinemaObject = dom.window.document.getElementsByClassName(
    '.showtime-box'
  );
  console.log(sfcinemaObject);
}, 11000);
```

But I'm getting the same result of zero nodes.

Or am I coding this wrong and need to do something in the constructor?

**Comment by domenic** [comment]
> HTMLCollection is an empty set.

This is false. Node.js just doesn't know how to log it correctly. Try logging `collection.length` or `collection[0]` instead.

**Comment by magician11** [comment]
```
const { JSDOM } = require('jsdom');

const dom = new JSDOM(``, {
  url: 'https://www.sfcinemacity.com/showtime/cinema/9936',
  runScripts: 'dangerously'
});

const sfcinemaObject = dom.window.document.getElementsByClassName(
  '.showtime-box'
);
console.log(sfcinemaObject.length);
console.log(sfcinemaObject[0]);
```

I get `0` and `undefined` respectively.

Can you show me the code snippet @domenic where you get something different?

**Comment by domenic** [comment]
No, in that case there really is a problem. It's just hard to diagnose when you're trying to draw conclusions from console.log'ing the HTMLCollection itself, since every HTMLCollection will look like `HTMLCollection {}`, even non-empty ones.

**Comment by TimothyGu** [comment]
@magician11 No your `new JSDOM(``, { url: '...' })` doesn't actually fetch the webpage. It merely sets the document URL. You still need `JSDOM.fromURL()`, which fetches the web page and returns a promise. Try

```js
const { JSDOM } = require('jsdom');

JSDOM.fromURL('https://www.sfcinemacity.com/showtime/cinema/9936', {
  runScripts: 'dangerously'
}).then(dom => {
  // ...
});
```

Then, you are using `document.getElementsByClassName` wrong. You don't need to prefix the class with a `.` for `getElementsByClassName`, but you do for `querySelector` or `querySelectorAll`.

----

**Now,** you need to apply what @Sebmaster mentioned: 

> Either by continually checking if the loader is still there, or you find some JS code on the page you can hook into.

Check [Asynchronous script loading](https://github.com/tmpvar/jsdom#asynchronous-script-loading) for some tips.

**Comment by magician11** [comment]
@TimothyGu good catch on the incorrectly used `.`.

I had a read of the asynchronous script loading docs... and the easiest start was to use a super long `setTimeout` as follows:

```
const { JSDOM } = require('jsdom');

JSDOM.fromURL('https://www.sfcinemacity.com/showtime/cinema/9936', {
  runScripts: 'dangerously'
}).then(dom => {
  setTimeout(() => {
    const sfcinemaObject = dom.window.document.getElementsByClassName(
      'showtime-box'
    );
    console.log(sfcinemaObject.length);
    console.log(sfcinemaObject[0]);
  }, 22000);
});
```
But I still get a result set of 0. Which is different from the Chrome console of 4 results. So not sure.. thoughts?

**Comment by rikku47** [comment]
Hi to all,

I have a similar problem. My Code:

```
let url = "anyUrl";
let selector = "anySelector";
let $ = require('jquery');
const jsdom = require("jsdom");
const { JSDOM } = jsdom;

const options = {
    runScripts: "dangerously",
    resources: "usable"
  };

  JSDOM.fromURL(url, options).then(dom => {
    dom.window.addEventListener("DOMContentLoaded", function(){
        let text = dom.window.document.querySelector(selector).textContent;
        $("#output").text(text);
      });
  });
```
If the content is dynamically, `textContent ` is `NULL`.

How can I let jsdom waiting for that?

Thanks.

Best regards rikku47

**Comment by ericminio** [comment]
Hi,
Perhaps this example can help:
[jsdom needs time to load external script](https://github.com/ericminio/learning-javascript/blob/master/school/tests/jsdom.spec.js)
Best,

**Comment by iboukris** [comment]
Hi, not sure if the same, but in my case I just wanted to be invoked after the 'load' event of the page itself has been called. This did the trick:
```javascript
dom.window.addEventListener('DOMContentLoaded', event => {
    dom.window.addEventListener('load', event => test(dom, event));
});

**Comment by domenic** [comment]
**Posted by GPT-6 Astra Medium**

There are several separate issues in the examples:

- `getElementsByClassName("showtime-box")` takes a class name without a leading dot.
- `new JSDOM("", { url: "..." })` creates an empty document with that URL; it does not fetch the page. Use `JSDOM.fromURL()` to fetch it.
- External scripts require both `runScripts: "dangerously"` and `resources: "usable"`. The later example using a 22-second timeout still omits `resources: "usable"`, so waiting longer does not enable those scripts.

Even with those options, `DOMContentLoaded` or `load` does not guarantee that application-specific asynchronous requests have finished. Wait for the application’s readiness signal or the required elements, as described in [Asynchronous script loading](https://github.com/jsdom/jsdom#asynchronous-script-loading).

Closing as a usage question. If a problem remains after correcting the setup, please provide a self-contained reproduction that isolates the missing behavior from the live website.
